internal/token/token_csp.go

```diff
@@ -0,0 +1,160 @@
+package token
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"io"
+	"log"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+)
+
+type CSPAuthorizeResponse struct {
+	IdToken      string `json:"id_token"`
+	TokenType    string `json:"token_type"`
+	ExpiresIn    int    `json:"expires_in"`
+	Scope        string `json:"scope"`
+	AccessToken  string `json:"access_token"`
+	RefreshToken string `json:"refresh_token"`
+}
+
+func (t *CspServerToServerTokenService) GetToken() string {
+
+	t.mutex.Lock()
+	defer t.mutex.Unlock()
+
+	if !t.tokenReady {
+		cspAccessToken, err := t.CallCSP()
+
+		if err != nil {
+			return "INVALID_TOKEN"
+		}
+
+		t.AccessToken = cspAccessToken
+		t.tokenReady = true
+	}
+
+	return t.AccessToken
+}
+
+func (t *CspServerToServerTokenService) CallCSP() (string, error) {
+	var oauthPath = "/csp/gateway/am/api/auth/authorize"
+	client := &http.Client{}
+
+	req, err := http.NewRequest("POST", t.CSPBaseUrl+oauthPath, strings.NewReader(url.Values{"grant_type": {"client_credentials"}}.Encode()))
+
+	if err != nil {
+		return "", err
+	}
+
+	req.Header.Add("Authorization", encodeCSPCredentials(t.CSPClientId, t.CSPClientSecret))
+	req.Header.Add("Content-Type", "application/x-www-form-urlencoded")
+
+	resp, err := client.Do(req)
+
+	if err != nil {
+		return "", err
+	}
+
+	defer resp.Body.Close()
+
+	body, err := io.ReadAll(resp.Body)
+
+	var cspResponse CSPAuthorizeResponse
+
+	err = json.Unmarshal(body, &cspResponse)
+
+	if !hasDirectIngestScope(cspResponse.Scope) {
+		log.Println("The CSP response did not find any scope matching 'aoa:directDataIngestion' which is required for Wavefront direct ingestion.")
```

Comment:
I like the added log statements, but I'm conflicted about creating unexpected log entries for customers and in our unit tests

Reply:
change: this should probably return an error